projections.go

```diff
@@ -0,0 +1,244 @@
+package eventsourcing
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/hallgren/eventsourcing/core"
+)
+
+type fetchFunc func() (core.Iterator, error)
+type callbackFunc func(e Event) error
+
+type ProjectionHandler struct {
+	register *Register // used to map the event types
+	Encoder  encoder
+	count    int
+}
+
+func NewProjectionHandler(register *Register, encoder encoder) *ProjectionHandler {
+	return &ProjectionHandler{
+		register: register,
+		Encoder:  encoder,
+	}
+}
+
+type Projection struct {
+	fetchF    fetchFunc
+	callbackF callbackFunc
+	handler   *ProjectionHandler
+	Pace      time.Duration // Pace is used when a projection is running and it reaches the end of the event stream
+	Strict    bool          // Strict indicate if the projection should return error if the event it fetches is not found in the regiter
+	Name      string
+}
+
+// Group runs projections concurrently
+type Group struct {
+	handler     *ProjectionHandler
+	projections []*Projection
+	cancelF     context.CancelFunc
+	wg          sync.WaitGroup
+	ErrChan     chan ProjectionResult
+}
+
+// ProjectionResult is the return type for a Group and Race
+type ProjectionResult struct {
+	Error            error
+	Name             string
+	LastHandledEvent Event
+}
+
+// Projection creates a projection that will run down an event stream
+func (ph *ProjectionHandler) Projection(fetchF fetchFunc, callbackF callbackFunc) *Projection {
+	projection := Projection{
+		fetchF:    fetchF,
+		callbackF: callbackF,
+		handler:   ph,
+		Pace:      time.Second * 10,            // Default pace 10 seconds
+		Strict:    true,                        // Default strict is active
+		Name:      fmt.Sprintf("%d", ph.count), // Default the name to it's creation index
+	}
+	ph.count++
+	return &projection
+}
+
+// Run runs the projection forever until the context is cancelled. When there are no more events to consume it
+// sleeps the set pace before it runs again.
+func (p *Projection) Run(ctx context.Context) ProjectionResult {
+	var result ProjectionResult
+	timer := time.NewTimer(0)
+	for {
+		select {
+		case <-ctx.Done():
+			if !timer.Stop() {
+				<-timer.C
+			}
+			return ProjectionResult{Error: ctx.Err(), Name: result.Name, LastHandledEvent: result.LastHandledEvent}
+		case <-timer.C:
+			result = p.RunToEnd(ctx)
+			if result.Error != nil {
+				return result
+			}
+		}
+		// rest
```

Comment:
Not sure if this comment is correct or needed at all?

I think the code here is pretty clear as is.
```suggestion

```